TDL15-CloudFormation-Jenkins/cf-deploy-jenkins.yaml

```diff
@@ -0,0 +1,245 @@
+          while true; do lsblk | grep xvdb && break; sleep 10; done
+          # Attach the additional EBS volume and mount the volume in /var/lib/Jenkins
+          sudo mkdir -p /var/lib/jenkins  
+          if [ -z ${SnapshotId} ] ;then echo -e "o\nn\np\n1\n\n\nw" | sudo fdisk /dev/sdb; sleep 3; sudo mkfs.ext4 /dev/sdb1; sudo e2label /dev/sdb1 JENKINS; fi
+          echo -e "LABEL=JENKINS     /var/lib/jenkins    ext4   defaults 0 0" >> /etc/fstab
+          mount -a
+          # Ensure that your software packages are up to date on your instance
+          sudo yum update –y
+          # install Jenkins
+          sudo yum install -y java-1.8.0-openjdk-devel
+          # Enable the Jenkins repository. Download the repo file and import the GPG key
+          sudo yum install -y wget
+          # sudo rpm -ivh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
+          sudo wget -O /etc/yum.repos.d/jenkins.repo http://pkg.jenkins-ci.org/redhat/jenkins.repo
+          sudo yum clean all
+          sudo rpm --import https://pkg.jenkins.io/redhat/jenkins.io.key
+          # make jenkins dirs:
+          sudo mkdir -p /home/jenkins /var/lib/jenkins/.ssh /var/cache/jenkins/war /var/log/jenkins
+          # Install the latest stable version of Jenkins
+          sudo yum install -y jenkins
+          # check whether Jenkins is running, type:
+          # systemctl status jenkins
+          # check jenkins login: 
+          sudo useradd -d /var/lib/jenkins jenkins # login dir
+          usermod -s /bin/bash jenkins # shell use dir
+          # set_jenkins_port  
+          sudo sed -i "s/JENKINS_PORT=\"8080\"/JENKINS_PORT=\"${JenkinsPort}\"/g" /etc/sysconfig/jenkins
+          # sshconfig: 
+          sudo echo -e "Host *\n\tStrictHostKeyChecking no\n" >> /var/lib/jenkins/.ssh/config
+          # ssh-keyscan -H hostname.com >> /var/lib/jenkins/.ssh/known_hosts
+          # Generate SSH key: 
+          if [ -z ${SnapshotId} ] ;then sudo ssh-keygen -t rsa -b 2048 -C jenkins@jenkins -N "" -f /var/lib/jenkins/.ssh/id_rsa; sudo cat /var/lib/jenkins/.ssh/id_rsa.pub > /var/lib/jenkins/.ssh/authorized_keys; fi
+          # set permissions
+          sudo chown -R jenkins:jenkins /home/jenkins /var/lib/jenkins /var/cache/jenkins /var/log/jenkins; sudo chmod 0700 /var/lib/jenkins/.ssh; sudo chmod 0600 /var/lib/jenkins/.ssh/*
+          # firewall: 
+          sudo firewall-cmd --permanent --add-service=ssh; sudo firewall-cmd --permanent --add-port='${JenkinsPort}'/tcp; sudo firewall-cmd --reload
+          # start the Jenkins service and enable it to start on system boot
+          sudo systemctl start jenkins
+          sudo systemctl enable jenkins
+          # install git
+          yum install -y git
+          # install AWS CLI
+          curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
+          unzip awscliv2.zip
+          sudo ./aws/install
```

Comment:
Userdata script is running as root, sudo is not required here.

Reply:
I know it don't need sudo, I test command lines on terminal first, so when I created userdata, I did not delete sudo.